### DISCUSSION TITLE
insert `img` inside of `h2`?

### DISCUSSION THREAD

**Discussion by deadcoder0904** [rehypejs/rehype#139] [Q&A]
i have a html block that looks like:

```html
<h2>...</h2>
<img>...</img>
```

i want the `img` to be inside of `h2` like:

```html
<h2>
  <img>...</img>
</h2>
```

the `h2` & `img` is inside mdx like:

```mdx
## this is headline

<Image pos="right" src="/gold-coins.png" alt="gold coins" />
```

i tried using:

```ts
import { visit } from 'unist-util-visit'
import { matches, select, selectAll } from 'unist-util-select'

import type { Node } from 'unist'
import type { Root, Parent, Heading, HTML } from 'mdast'

export const rehypeHeadingImage = (options = {}) => {
	console.log('rehypeHeadingImage')
	return (tree: Root) => {
		console.log(tree)
		const x = (node: any) => node.type === 'element' || node.type === 'mdxJsxFlowElement' // the `img` above is `next/image` which has the type `mdxJsxFlowElement`
		const s = selectAll('heading', tree)
		console.log({ s, x: select('div', tree) })
		visit(tree, x, visitor)
	}
...
```

but for some reason, i'm unable to read selectors from `select` & `selectAll`... they both come empty `[]`. my actual selector is `h2 > img` but i can't get it to work.

so i gave that up after trying 2 days every which way.

then now i've recently downloaded `rehype-rewrite` but unable to figure out, how to put the node inside:

```ts
[
					rehypeRewrite,
					{
						rewrite(node: any, index: number, parent: any) {
							if (node.tagName === 'h2' || node.tagName === 'h3' || node.tagName === 'h4') {
								console.log(node.children[0].value)
								console.log(node)
								// node.children = [...node.children]
                                // idk how to get sibling selector from here so i can remove it & put it inside h2
							}
						},
					},
				],
```

no matter how many times i work with `unified` & `rehype`, it still confuses me.

i did read the explore & learn again but no luck.

**edit:** i found a solution but definitely don't prefer it. insert `img` from inside `rehype-rewrite` by calculating `h2`'s so every image is in array `[img1, img2]` & they get inserted but this requires some calculation as i don't add image after every `h2` but only some of them... which can be calculated too but still not preferred if i can do it in an easy way.

**Comment by wooorm** [comment]
You can remove most lines of your code, they are not needed. For example, you are not dealing with mdast, so you don‘t have to import those types. And, you don’t need all those tools. You especially don’t need `rehype-rewrite`.

You also use TypeScript but cast things to `any`: if you don‘t want types, don’t use TypeScript. If you do, actually use it.

Instead, I recommend using the visitor pattern to traverse the tree: https://unifiedjs.com/learn/recipe/tree-traversal/. It can do everything you need, is fast, and is simple. To illustrate:

```js
import {visit} from 'unist-util-visit'

/** @type {import('hast').Root} */
const tree = {}

visit(tree, function (node, index, parent) {
  if (
    node.type === 'element' &&
    node.tagName === 'h2' &&
    typeof index === 'number' &&
    parent
  ) {
    const next = parent.children[index + 1]

    if (next && next.type === 'element' && next.tagName === 'img') {
      node.children.push(next)
      parent.children.splice(index + 1, 1)
    }
  }
})
```

**Reply by wooorm** [reply]
I missed that you are actually dealing with MDX. Not plain rehype. So it’s not an image element, but apparently a Next.js `Image` component? In that case, please change the condition for `if (next ...`.

**Reply by deadcoder0904** [reply]
help me understand bcz it didn't work.

the image is `next/image` but its wrapped in a `div` like:

```ts
return (
	<div
		className={clsx('pic absolute', className, {
			'left-32': pos === 'left',
			'right-32': pos === 'right',
		})}
	>
		{/* eslint-disable-next-line @next/next/no-img-element */}
		<NextImage
			className="relative hidden lg:block"
			src={src}
			alt={alt}
			width={296}
			height={296}
		/>
	</div>
)
```

i changed the code to this:

```ts
import { visit } from 'unist-util-visit'

import type { Root, Parent } from 'hast'

export const rehypeHeadingImage = (options = {}) => {
	console.log('rehypeHeadingImage')
	return (tree: Root) => {
		visit(tree, function (node, index, parent) {
			if (node.type === 'element' && node.tagName === 'h2' && typeof index === 'number' && parent) {
				const next = parent.children[index + 2]
				console.log({ next })
				if (next && next.type === 'element' && next.tagName === 'div') {
					node.children.push(next)
					console.log({ node })
					parent.children.splice(index + 2, 1)
				}
			}
		})
	}
}
```

but for some reason, the html structure didn't change.

when i log `next` i get the following console log:

```ts
{
  next: {
    type: 'mdxJsxFlowElement',
    name: 'Image',
    attributes: [ [Object], [Object], [Object] ],
    children: [],
    position: { start: [Object], end: [Object] },
    data: { _mdxExplicitJsx: true }
  }
}
```

but if i try `next.type === 'mdxJsxFlowElement'` or `next.name === 'Image'`, i get the following errors respectively:

> This comparison appears to be unintentional because the types '"comment" | "doctype" | "element" | "text" | "raw"' and '"mdxJsxFlowElement"' have no overlap.ts(2367)
(property) type: "comment" | "doctype" | "element" | "text" | "raw"

and

> Property 'name' does not exist on type 'Comment | DocType | Element | Text | Raw'.
  Property 'name' does not exist on type 'Comment'.

i'm also not sure if i have to specify `node.children = node` & `parent.children = parent` or to equate `tree` to something or it will mutate the DOM bcz of `.push()`

**Reply by wooorm** [reply]
Please show your exact input MDX file.

You are now showing JSX code (with the div) that is used at runtime. It is not in an MDX file. The AST does not include access to other files you may have on your computer. MDX does not read other files on your computer.

Why do you use `index + 2`?

Do you deal with **HTML** or with **markdown**?

If you want to look at the AST, I recommend using a `console.log` statement. Particularly, I like `console.dir`, which allows you to print deep objects:

```
console.dir(tree, {depth: 10}) // print 10 levels deep.
```

>  i get the following errors respectively:

MDX syntax is not normally in HTML files. So, MDX nodes are not normally in hast types. You need to register types. See for a solution: https://github.com/syntax-tree/mdast-util-mdx#types

> i'm also not sure if i have to specify node.children = node & parent.children = parent or to equate tree to something or it will mutate the DOM bcz of .push()


None of these, they are all incorrect.

**Reply by deadcoder0904** [reply]
my `.mdx` file looks like:

```mdx
---
---

## this is the headline

<Image pos="right" src="/gold-coins.png" alt="gold coins" />

lots of paragraphs
```

> Why do you use index + 2?

bcz it returned nothing as it was showing `type: text` with `\n` as a value so i had to use `index + 2`

> Do you deal with HTML or with markdown?

i deal with `mdx` as an input for a landing page. that turns into `html`... `mdx` contains both text & images.

im using it like a blog layout. mdx -> mdx-bundler -> html

> None of these, they are all incorrect.

so will the html modify itself? bcz it isn't happening currently.

**Reply by wooorm** [reply]
> i deal with mdx as an input for a landing page. that turns into html... mdx contains both text & images.

> im using it like a blog layout. mdx -> mdx-bundler -> html

Are you using the plugin you are writing **inside** MDX? Or *after* MDX?
In other words, do you pass it to `@mdx-js/mdx` or one of its wrappers (such as `mdx-bundler`), or do you use `rehype` and friends on the final HTML file?

> so will the html modify itself? bcz it isn't happening currently.

Because I don’t yet have enough information to exactly solve all your code for you.

**Reply by deadcoder0904** [reply]
oh i use it like this in `mdx.ts`:

```ts
import fs from 'fs/promises'
import path from 'path'
import { bundleMDX } from 'mdx-bundler'
import remarkGfm from 'remark-gfm'

import { rehypeHeadingImage } from '~/rehype/heading-image'

import type { Meta, Slug } from '~/types/index'

const ROOT_PATH = process.cwd()

export const DATA_PATH = path.join(ROOT_PATH, 'src/_ab-testing')

const PUBLIC_PATH = path.join(ROOT_PATH, 'public')

interface IGetPostBySlug {
	meta: Meta
	code: string
	folderPath: string
}

export const getPostBySlug = async (slug: Slug): Promise<IGetPostBySlug> => {
	const source = await fs.readFile(path.join(DATA_PATH, slug as string, 'index.mdx'), 'utf8')

	if (process.platform === 'win32') {
		process.env.ESBUILD_BINARY_PATH = path.join(
			process.cwd(),
			'node_modules',
			'esbuild',
			'esbuild.exe'
		)
	} else {
		process.env.ESBUILD_BINARY_PATH = path.join(
			process.cwd(),
			'node_modules',
			'esbuild',
			'bin',
			'esbuild'
		)
	}

	const cwd = path.join(DATA_PATH, slug)

	const { code, frontmatter } = await bundleMDX({
		source,
		cwd,
		mdxOptions: (options, frontmatter) => {
			options.rehypePlugins = [
				...(options.rehypePlugins ?? []),
				// rehypeMetaAttribute,
				rehypeHeadingImage,
				// rehypeSlug,
				// [rehypeAutolinkHeadings, { properties: { className: 'anchor' } }],
				// rehypeAccessibleEmojis,
			]
			options.remarkPlugins = [
				...(options?.remarkPlugins ?? []),
				remarkGfm,
				// [remarkTocHeadings, { exportRef: toc }],
			]
			return options
		},
		esbuildOptions: (options) => {
			options.outdir = path.join(PUBLIC_PATH, slug)
			options.loader = {
				...options.loader,
				'.webp': 'file',
				'.jpeg': 'file',
				'.jpg': 'file',
				'.png': 'file',
				'.gif': 'file',
			}
			options.publicPath = slug

			options.write = true

			return options
		},
	})

	const meta = {
		...frontmatter,
		slug,
	} as Meta

	return {
		meta,
		code,
		folderPath: cwd,
	}
}
```

its basically this repo → https://github.com/deadcoder0904/mdx-bundler-gif-png-error/blob/main/src/utils/mdx.ts (similar `mdx.ts`)

**Reply by wooorm** [reply]
~~Common, you show an import of something that is never used? Do you not use this plugin?~~ Oh wait you show the whole file, missed that

I know you know how this works: I am going to ask you 100s of questions because you don’t provide enough info. Please please please please provide info...

**Reply by deadcoder0904** [reply]
i just showed `rehypeHeadingImage` which you wrote. you literally asked for if its used before or after mdx so i showed exactly that file.

yeah idk what to answer exactly.

so i'll repeat what you asked previously:

i use `mdx-bundler` with a simple blog layout. this is similar to mine → https://www.peterlunch.com/blog/mdx-bundler-beginners

i dont use `@mdx-js/mdx`... i use `@next/mdx`... i bootstrapped the project through `create-trpc`

this is my `next.config.mjs`:

```mjs
/**
 * Run `build` or `dev` with `SKIP_ENV_VALIDATION` to skip env validation. This is especially useful
 * for Docker builds.
 */
await import('./src/env.mjs')
import createMDX from '@next/mdx'

const withMDX = createMDX({
	extension: /\.mdx?$/,
	options: {
		// If you use remark-gfm, you'll need to use next.config.mjs
		// as the package is ESM only
		// https://github.com/remarkjs/remark-gfm#install
		remarkPlugins: [],
		rehypePlugins: [],
		// If you use `MDXProvider`, uncomment the following line.
		providerImportSource: '@mdx-js/react',
	},
})

/** @type {import("next").NextConfig} */
const config = {
	reactStrictMode: true,
	// Configure pageExtensions to include md and mdx
	pageExtensions: ['ts', 'tsx', 'js', 'jsx', 'md', 'mdx'],
	/**
	 * If you have `experimental: { appDir: true }` set, then you must comment the below `i18n` config
	 * out.
	 *
	 * @see https://github.com/vercel/next.js/issues/41980
	 */
	i18n: {
		locales: ['en'],
		defaultLocale: 'en',
	},
}

export default withMDX(config)
```

my entire goal is to place `img` inside `h2`. its technically a `next/image` but the component looks like `div > next/image` bcz of styling.

let me know if there's anything i should answer or create a small repro.

**Reply by wooorm** [reply]
Sorry, all this back and forth sometimes gets on my nerves.

First of all: why do you even want this? Move the components into the headings! Much simpler. Particularly if you are going to spend days on trying to automate this: just do it manually.

Ok, then. You don’t need to do this on the HTML level. The markdown level is a bit simpler. In HTML, you have stuff like whitespace. In markdown, you don’t need to deal with that.

So, here’s a remark plugin, which support MDX nodes, and is a version of my earlier pseudo-code:

```js
/**
 * @typedef {import('remark-mdx')}
 */

import {visit} from 'unist-util-visit'

/** @type {import('unified').Plugin<[], import('mdast').Root>} */
export default function myRemarkPluginAddingComp() {
  return function (tree) {
    visit(tree, function (node, index, parent) {
      if (
        node.type === 'heading' &&
        node.depth === 2 &&
        typeof index === 'number' &&
        parent
      ) {
        const next = parent.children[index + 1]

        if (
          next &&
          next.type === 'mdxJsxFlowElement' &&
          next.name === 'Image'
        ) {
          node.children.push(next)
          parent.children.splice(index + 1, 1)
        }
      }
    })
  }
}
```

**Reply by wooorm** [reply]
> you literally asked for if its used before or after mdx so i showed exactly that file.

I didn’t expent an entire file being dumped. I expected a “yes, I pass plugins in `mdxOptions` to `mdx-bundler`”.

> `@next/mdx`

That project is a) a *thin* wrapper around `@mdx-js/mdx`, and b) an *alternative* to `mdx-bundler`.
Why do you use both `mdx-bundler` and an alternative to it?

> my entire goal is to place img inside h2. its technically a next/image but the component looks like div > next/image bcz of styling.

OK, good to hear your *practical goal on the frontend side* (an `img` in an `h2`).
It is unrelated that those are `next/image`s with `div`s.
What is related is what your MDX files look like.

**Reply by deadcoder0904** [reply]
> I didn’t expent an entire file being dumped. I expected a “yes, I pass plugins in `mdxOptions` to `mdx-bundler`”.

yeah i did it bcz half the time i don't understand `unified` stuff & your questions so i dumped it bcz it was only like <100 locs so its very easy to read bcz u understand that stuff.

anyways, thank you so much. idk what would happen to unified ecosystem tho bcz only 1 guy has all the knowledge. and majority of it cant be googled.

i'll probably do this manually if this doesn't work. just plain ol' html.

thanks again & dont worry you aint that bad. otoh, you are too patient 😊

**Reply by deadcoder0904** [reply]
it didn't work so i'll drop that idea. it was stupid anyways. going back to html :))

**Reply by wooorm** [reply]
> idk what would happen to unified ecosystem tho bcz only 1 guy has all the knowledge. and majority of it cant be googled.

Have you ever written a babel plugin? PostCSS? SWC? It‘s complex programming, it isn’t super easy. But the entire frontend/JS stack is build on ASTs and plugins.

Again: it is not easy, but a bunch of folks have the knowledge; you can get there too, with a bit more patience, focussing on cleaner code and understanding what things do, and a couple more years of experience.

**Reply by deadcoder0904** [reply]
nope. i haven't even coded in a long while either so my skills have gotten rusty too.

my focus is on being an entrepreneur rather than a programmer so when i make enough money to hire best programmers in the world like you, i'll prolly quit.

this stuff is definitely a different domain than my usual stuff but its basically data structures sprinkled with dom manipulation. did some ast back in 2016-17 for college.

anyways, thanks for your help again. i started with html. i just didnt wanna do that bcz huge html. then i saw tailwindcss landing page & its huge as well so i did that & its much easier to css that.
